### PR TITLE
Garbage collection per token range

### PR DESCRIPTION
**Pull request by mlaflamm**

### CODE REVIEW

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala [hunk 1]
```diff
@@ -292,39 +275,84 @@ class MysqlStorage(config: MysqlStorageConfiguration, garbageCollection: Boolean
...
+
     private[mysql] def kill() {
       this.stop()
       this.scheduledTask.cancel(true)
     }
 
     private[mysql] def collectAll(toCollect: Int): Int = {
-      allTables.foldLeft(0)((sum, table) => sum + collect(table, toCollect))
+      val collectors = tableCollectors.synchronized {
+        allTables.map(table => getTableCollector(table)).toList
```
Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala]: Is .toList usefull?

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala [hunk 2]
```diff
@@ -292,39 +275,84 @@ class MysqlStorage(config: MysqlStorageConfiguration, garbageCollection: Boolean
...
+    }
+  }
+
+  class TableCollector(table: Table, tokenRanges: List[TokenRange]) {
+    private val metricCollect = metrics.timer("gc-collect", table.uniqueName)
+    private val metricCollected = metrics.meter("gc-collected-record", "records", table.uniqueName)
+
+    private var tableNextRange: TokenRange = tokenRanges.head
+    private var tableNextToken: Long = tokenRanges.head.start
+    private var tableVersionsCache = mutable.Queue[VersionRecord]()
```
Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala]: Mutable Queue in a var?

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala [hunk 3]
```diff
@@ -348,14 +375,16 @@ class MysqlStorage(config: MysqlStorageConfiguration, garbageCollection: Boolean
               nextToken = record.token
           }
 
-          if (nextToken >= Int.MaxValue.toLong * 2)
-            nextToken = 0
-
-          tableNextToken += (table -> nextToken)
-          tableVersionsCache += (table -> recordsVersions)
+          if (tableVersionsCache.size == 0 && nextToken >= lastRange.end) {
+            tableNextRange = lastRange.nextRange(tokenRanges).getOrElse(tokenRanges.head)
```
Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala:392]: What if tokenRanges is empty?

Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala:392]: Good catch! I modified the code to never create a TableCollector if the range is empty...

**Comment by alexbergeron** [comment]
+1

**Pull request merged**